Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,337 @@
+      RecursiveIn g (fun p =>
+        let (a, n) := Nat.unpair p
+        n.rec (f a) (fun y ih => do
+          let i ← ih
+          h (Nat.pair a (Nat.pair y i))))
+  | rfind {f : ℕ →. ℕ} (hf : RecursiveIn g f) :
+      RecursiveIn g (fun a =>
+        Nat.rfind (fun n => (fun m => m = 0) <$> f (Nat.pair a n)))
+
+/--
+`f` is Turing reducible to `g` if `f` is recursive in `g`.
+-/
+def turing_reducible (f g : ℕ →. ℕ) : Prop :=
+  RecursiveIn g f
+
+/--
+Custom infix notation for `turing_reducible`.
+-/
+infix:50 " ≤ᵀ " => turing_reducible
+
+/--
+`f` is Turing equivalent to `g` if `f` is reducible to `g` and `g` is reducible to `f`.
+-/
+def turing_equivalent (f g : ℕ →. ℕ) : Prop :=
+  f ≤ᵀ g ∧ g ≤ᵀ f
+
+/--
+Custom infix notation for `turing_equivalent`.
+-/
+infix:50 " ≡ᵀ " => turing_equivalent
+
+/--
+A partial function `f` is partial recursive if and only if
+it is recursive in the constant zero function.
+-/
+lemma partrec_iff_partrec_in_zero
+  (f : ℕ →. ℕ) : Nat.Partrec f ↔ RecursiveIn (fun _ => pure 0) f := by
+  constructor
+  { intro pF
+    induction' pF
+    case mp.zero =>
+      apply RecursiveIn.zero
+    case mp.succ =>
+      apply RecursiveIn.succ
+    case mp.left =>
+      apply RecursiveIn.left
+    case mp.right =>
+      apply RecursiveIn.right
+    case mp.pair _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.pair ih1 ih2
+    case mp.comp _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.comp ih1 ih2
+    case mp.prec _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.prec ih1 ih2
+    case mp.rfind _ _ ih =>
+      apply RecursiveIn.rfind ih }
+  { intro fRecInNone
+    induction' fRecInNone
+    case mpr.zero =>
+      apply Nat.Partrec.zero
+    case mpr.succ =>
+      apply Nat.Partrec.succ
+    case mpr.left =>
+      apply Nat.Partrec.left
+    case mpr.right =>
+      apply Nat.Partrec.right
+    case mpr.oracle =>
+      apply Nat.Partrec.zero
+    case mpr.pair _ _ _ _ ih1 ih2 =>
+      apply Nat.Partrec.pair ih1 ih2
+    case mpr.comp _ _ _ _ ih1 ih2 =>
+      apply Nat.Partrec.comp ih1 ih2
+    case mpr.prec _ _ _ _ ih1 ih2 =>
+      apply Nat.Partrec.prec ih1 ih2
+    case mpr.rfind _ _ ih =>
+      apply Nat.Partrec.rfind ih }
+
+/--
+A partial function `f` is partial recursive if and only if it is recursive in
+every partial function `g`.
+-/
+theorem partrec_iff_partrec_in_everything
+  (f : ℕ →. ℕ) : Nat.Partrec f ↔ (∀ g, RecursiveIn g f) := by
+  constructor
+  { intro pF
+    intro g
+    induction pF
+    case zero =>
+      apply RecursiveIn.zero
+    case succ =>
+      apply RecursiveIn.succ
+    case left =>
+      apply RecursiveIn.left
+    case right =>
+      apply RecursiveIn.right
+    case pair _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.pair ih1 ih2
+    case comp _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.comp ih1 ih2
+    case prec _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.prec ih1 ih2
+    case rfind _ _ ih =>
+      apply RecursiveIn.rfind ih }
```

Comment:
This duplicates `partrec_iff_partrec_in_zero`, perhaps first prove `Nat.Partrec f -> ∀ g, RecursiveIn g f` and `RecursiveIn (fun _ => pure 0) f -> Nat.Partrec f` and conclude both iffs from these two implications?